hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/BucketUploadManager.java

```diff
@@ -0,0 +1,197 @@
+        } catch (IOException e) {
+            logger.error("Failed to create block stream uploader directories: {}", e.getMessage());
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void onBucketConfigurationsUpdated(@NonNull List<CompleteBucketConfig> configs) {
+        // Acquire write lock to update uploaders
+        uploadersLock.writeLock().lock();
+        try {
+            // Clear all existing uploaders
+            uploaders.clear();
+
+            // Create new uploaders for each config
+            for (CompleteBucketConfig config : configs) {
+                if (config.enabled()) {
+                    try {
+                        CloudBucketUploader newUploader =
+                                new MinioBucketUploader(executorService, configProvider, config);
+                        uploaders.add(newUploader);
+                        logger.info("Created new uploader for bucket: {}", config.bucketName());
+                    } catch (Exception e) {
+                        logger.error(
+                                "Failed to create uploader for bucket {}: {}", config.bucketName(), e.getMessage());
+                    }
+                }
+            }
+        } finally {
+            uploadersLock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public void onBlockClosed(@NonNull Path blockPath) {
+        // Start a new async task for handling the block closure
+        CompletableFuture<Void> uploadTask = CompletableFuture.runAsync(
+                () -> {
+                    try {
+                        processBlockClosure(blockPath);
+                    } catch (Exception e) {
+                        logger.error("Error processing block closure for {}: {}", blockPath, e.getMessage());
+                    }
+                },
+                executorService);
+
+        // Store the task for tracking
+        pendingUploads.put(blockPath, uploadTask);
+
+        // Clean up the tracking when the task completes
+        uploadTask.whenComplete((result, error) -> pendingUploads.remove(blockPath));
+    }
+
+    private void processBlockClosure(Path blockPath) {
+        // Acquire read lock to access uploaders
+        uploadersLock.readLock().lock();
+        try {
+            if (uploaders.isEmpty()) {
+                logger.warn("No cloud storage providers registered for upload");
+                return;
+            }
+
+            // Create upload tasks for each provider
+            List<CompletableFuture<Void>> uploadTasks = uploaders.stream()
+                    .map(uploader -> uploadToProvider(blockPath, uploader))
+                    .toList();
+
+            // Wait for all uploads to complete and handle the results
+            CompletableFuture.allOf(uploadTasks.toArray(CompletableFuture[]::new))
+                    .thenAccept(v -> handleUploadCompletion(blockPath))
+                    .exceptionally(throwable -> {
```

Comment:
Having `.exceptionally()` here will handle only one of the possible exceptions in `uploadTasks` right?
If that's the case we can perhaps have the `.exceptionally()` for each CompletableFuture in `uploadToProvider()` so that we can handle each exception separately e.g. when we need to update `uploads.failure` metric per provider.